steward/src/cellars/aave_v2_stablecoin.rs

```diff
@@ -1,77 +1,182 @@
+            Ok(AaveV2StablecoinCellarCalls::TransferFees(call).encode())
         }
-        SetInputToken(params) => {
-            let call = SetInputTokenCall {
-                token: params
-                    .address
-                    .parse::<H160>()
-                    .expect("failed to parse token address"),
-                is_approved: params.is_approved,
-            };
-            AaveV2StablecoinCellarCalls::SetInputToken(call).encode()
+    }
+}
+
+fn validate_route(results: Vec<Result<H160, &String>>) -> Result<(), Error> {
+    let mut bad_addresses = String::new();
+    for r in results {
+        if let Err(addr) = r {
+            bad_addresses.push_str(&format!(", {}", addr))
         }
-        RemoveLiquidityRestriction(_) => {
-            let call = RemoveLiquidityRestrictionCall {};
-            AaveV2StablecoinCellarCalls::RemoveLiquidityRestriction(call).encode()
+    }
+
+    if !bad_addresses.is_empty() {
+        let mut err_string = "Rebalance 'route': array contains invalid address(s)".to_string();
+        err_string.push_str(&bad_addresses);
+        return Err(sp_call_error(format!("{}: {}", LOG_PREFIX, err_string)));
+    }
+
+    Ok(())
+}
+
+fn validate_swap_params(swap_params: Vec<SwapParams>) -> Result<(), Error> {
+    for sp in swap_params.iter() {
+        if sp.in_index == 0u64 && sp.out_index == 0u64 && sp.swap_type == 0u64 {
+            continue;
         }
-        Sweep(params) => {
-            let call = SweepCall {
-                token: params
-                    .address
-                    .parse::<H160>()
-                    .expect("failed to parse token address"),
-            };
-            AaveV2StablecoinCellarCalls::Sweep(call).encode()
+
+        if sp.out_index > 8u64 {
+            return Err(sp_call_error(format!(
+                "{}: Rebalance 'swap_params': out_index can't be greater than 8 (the last index of 'route'). {:?}.",
+                LOG_PREFIX,
+                sp
+            )));
+        }
+
+        if sp.in_index % 2 != 0u64 || sp.out_index % 2 != 0u64 {
+            return Err(sp_call_error(format!(
+                "{}: Rebalance 'swap_params': token indeces must be even. {:?}.",
+                LOG_PREFIX, sp
+            )));
+        }
+
+        let diff = sp.out_index.checked_sub(sp.in_index);
+        if diff.is_none() || diff.unwrap() == 0u64 {
+            return Err(sp_call_error(format!(
+                "{}: Rebalance 'swap_params': in_index must be less than out_index. {:?}.",
+                LOG_PREFIX, sp
+            )));
+        }
+
+        let diff = diff.unwrap();
+        if diff != 2 {
+            return Err(sp_call_error(format!(
+                "{}: Rebalance 'swap_params': in_index and out_index must surround a pool address (must 2 apart). {:?}.",
+                LOG_PREFIX,
+                sp
+            )));
         }
     }
+
+    Ok(())
+}
+
+fn sp_call_error(message: String) -> Error {
```

Comment:
We've defined this function here in a cellar-specific file, but we also manually call `ErrorKind::SPCallError.context().into()` elsewhere, do we want this helped function moved and more broadly used?

Reply:
Yes but I figured it would need to be a method per error variant otherwise it becomes almost as verbose as what it's replacing. I do think it would make more sense to house it in `error.rs` or `utils.rs`.